### PR TITLE
Update 'initialize member from parameter' to work with primary constructors.

### LINKED ISSUE #68817: 'Create and assign [field|property]' does not work for primary constructor parameters
This refactoring is offered for constructor parameters:
image

It should also be offered for primary constructor parameters.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/68817

Looks like this:

image

QOL for people using primary constructors.  Just like witha  normal constructor, we want to make it easy to be writing the constructor and then have the user go and create fields/props easily.

Fixes #68817

### CODE REVIEW

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromParameterCodeRefactoringProvider.cs [hunk 1]
```diff
@@ -56,55 +56,10 @@ protected override SyntaxNode GetBody(SyntaxNode functionDeclaration)
             => InitializeParameterHelpers.GetBody(functionDeclaration);
 
         protected override SyntaxNode? GetAccessorBody(IMethodSymbol accessor, CancellationToken cancellationToken)
-        {
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromParameterCodeRefactoringProvider.cs:59]: moved to common helper locations.

File: src/Features/Core/Portable/InitializeParameter/AbstractInitializeMemberFromParameterCodeRefactoringProviderMemberCreation.cs [hunk 2]
```diff
@@ -314,19 +316,6 @@ internal abstract partial class AbstractInitializeMemberFromParameterCodeRefacto
             throw ExceptionUtilities.Unreachable();
         }
 
-        private static string GenerateUniqueName(IParameterSymbol parameter, ImmutableArray<string> parameterNameParts, NamingRule rule)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/AbstractInitializeMemberFromParameterCodeRefactoringProviderMemberCreation.cs:317]: moved to common helper location.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractInitializeMemberFromParameterCodeRefactoringProviderMemberCreation.cs:317]: same with the other deletions in the file

File: src/Features/Core/Portable/InitializeParameter/InitializeParameterHelpersCore.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+using Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.InitializeParameter
+{
+    internal static class InitializeParameterHelpersCore
+    {
+        public static ImmutableArray<(IParameterSymbol parameter, bool before)> GetSiblingParameters(IParameterSymbol parameter)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/InitializeParameter/InitializeParameterHelpersCore.cs:15]: these are just moves of helpers.

File: src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs [hunk 4]
```diff
@@ -138,5 +142,86 @@ public static bool IsExpressionBody(SyntaxNode body)
                 _ => throw ExceptionUtilities.UnexpectedValue(body),
             };
         }
+
+        public static SyntaxNode? GetAccessorBody(IMethodSymbol accessor, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InitializeParameter/InitializeParameterHelpers.cs:146]: these are just moves of helpers.

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -0,0 +1,706 @@
...
+                return;
+
+            if (selectedParameter.Parent is not ParameterListSyntax { Parent: TypeDeclarationSyntax(kind: SyntaxKind.ClassDeclaration or SyntaxKind.StructDeclaration) typeDeclaration })
+                return;
+
+            // var parameterNodes = generator.GetParameters(functionDeclaration);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            // we can't just call GetDeclaredSymbol on functionDeclaration because it could an anonymous function,
+            // so first we have to get the parameter symbol and then its containing method symbol
```
Comment (sharwell) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs]: 💡 This comment seems specific to non-primary-constructor scenarios

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs [hunk 6]
```diff
@@ -0,0 +1,706 @@
...
+            // textSpan and use that as `parameterNodes`, where `selectedParameter` would be the first one.
+
+            var selectedParameter = await context.TryGetRelevantNodeAsync<ParameterSyntax>().ConfigureAwait(false);
+            if (selectedParameter == null)
+                return;
+
+            if (selectedParameter.Parent is not ParameterListSyntax { Parent: TypeDeclarationSyntax(kind: SyntaxKind.ClassDeclaration or SyntaxKind.StructDeclaration) typeDeclaration })
+                return;
+
+            // var parameterNodes = generator.GetParameters(functionDeclaration);
```
Comment (sharwell) [inline src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs]: ❔ Still relevant?

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs [hunk 7]
```diff
@@ -198,7 +200,7 @@ private static (TParameterSyntax?, IParameterSymbol?) GetParameterAtOrdinal(int
                 var operation = semanticModel.GetOperation(coalesceNode, cancellationToken);
                 if (operation is ICoalesceOperation coalesceExpression)
                 {
-                    if (IsParameterReference(coalesceExpression.Value, parameter) &&
+                    if (InitializeParameterHelpersCore.IsParameterReference(coalesceExpression.Value, parameter) &&
```
Comment (sharwell) [inline src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs:203]: ❔ Is this change still necessary with the `using static` above?

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs]: removed!

File: src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/InitializeParameter/CSharpInitializeMemberFromPrimaryConstructorParameterCodeRefactoringProvider.cs]: removed!

File: src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/InitializeParameter/AbstractAddParameterCheckCodeRefactoringProvider.cs]: nope.  cleaned.

File: src/Analyzers/CSharp/CodeFixes/AssignOutParameters/AbstractAssignOutParametersCodeFixProvider.cs [hunk 8]
```diff
@@ -110,10 +110,10 @@ private static bool IsValidLocation(SyntaxNode location)
...
                 Contract.ThrowIfNull(parameterList);
 
-                var outParameters =
-                    parameterList.Parameters.Select(p => (IParameterSymbol)semanticModel.GetRequiredDeclaredSymbol(p, cancellationToken))
-                                            .Where(p => p.RefKind == RefKind.Out)
-                                            .ToImmutableArray();
+                var outParameters = parameterList.Parameters
+                    .Select(p => semanticModel.GetRequiredDeclaredSymbol(p, cancellationToken))
+                    .Where(p => p.RefKind == RefKind.Out)
+                    .ToImmutableArray();
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/AssignOutParameters/AbstractAssignOutParametersCodeFixProvider.cs:116]: cast removed as this now uses a typesafe extension.

**Review by akhera99 — APPROVED** [review]
